### PR TITLE
Enable -Werror=format compile errors on torch exception types

### LINKED ISSUE #33899: Segfault upon calling torch.Tensor on a gpu tensor
## 🐛 Bug

I accidentally called `torch.Tensor` on a Tensor object, thinking it was a numpy array. Instead of doing nothing or throwing a warning, the program segfaulted. 

## To Reproduce

Minimal code example:

```python
import torch
import numpy as np
device = torch.device('cuda')
a = torch.Tensor(np.random.normal(size=10)).to(device)
torch.Tensor(a)
```

Upon running the final line, python prints "Segmentation fault (core dumped)" and dies.

## Expected behavior

I would have expected casting a Tensor to a Tensor to have to result

## Environment

Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2060
Nvidia driver version: 430.50
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] numpydoc==0.9.1
[pip] torch==1.3.1
[pip] torchfile==0.1.0
[pip] torchvision==0.4.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.0.2            py37h7b6447c_0  
[conda] mkl_fft                   1.0.12           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.4.2                    pypi_0    pypi

## Additional context

I was able to reproduce the issue running on google CoLab, so I don't think the issue is specific to my computer

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #33899

In the issue, we have
```
TypeError("expected %s (got %s)", dispatch_key, toString(other.key_set()).c_str());
```
which results in `dispatch_key` being interpreted as a c-string by `sprintf`. Adding `__attrbute__((format))` to the `TypeError` constructor allows gcc or clang to detect this at compile time. Then `-Werror=format` makes it a hard error at compile time.

### CODE REVIEW

File: torch/csrc/utils/tensor_numpy.cpp [hunk 1]
```diff
@@ -76,13 +76,13 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
...
-      "copy the tensor to host memory first.");
+      "can't convert %s device type tensor to numpy. Use Tensor.cpu() to "
+      "copy the tensor to host memory first.", tensor.device().str().c_str());
   }
   if (tensor.layout() != Layout::Strided) {
       throw TypeError(
-        "can't convert non-strided tensor to numpy."
-        "convert the tensor to a strided layout first.");
+        "can't convert %s layout tensor to numpy."
+        "convert the tensor to a strided layout first.", c10::str(tensor.layout()).c_str());
```
Comment (peterbell10) [inline torch/csrc/utils/tensor_numpy.cpp]: This was a format error on the commit I was working on but rebasing shows this was already fixed in #33612. This does give a more informative error though.

cc @mruberry

**Comment by peterbell10** [comment]
Had a look into the CI failures:
- XLA test needs #33908 which isn't available on `viable/strict` yet
- BC check is broken on [other PRs](https://circleci.com/gh/pytorch/pytorch/4648407?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) and looks to be related to named tensors which I haven't touched here.

File: torch/csrc/utils/tensor_numpy.cpp (hunk 1 above)
Comment (mruberry) [reply torch/csrc/utils/tensor_numpy.cpp]: SGTM.

**Review by ezyang — APPROVED** [review]
It would be better to stop using format strings entirely, but this is good too.

**Comment by peterbell10** [comment]
I had started to rewrite it using `c10::str` but it occured to me that this might be a part of `torch`'s C++ API. If that's not an issue I can make that change in a follow-up PR.

**Comment by ezyang** [comment]
I... don't think TypeError is public API. It doesn't matter too much, up to you :)

**Pull request closed without merging**